modules/key-distributor/cfn/encrypter/function.go

```diff
@@ -0,0 +1,163 @@
+/*
+Copyright 2020 Google LLC
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    https://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package encrypter
+
+import (
+	"bytes"
+	"context"
+	"encoding/base64"
+	"encoding/json"
+	"fmt"
+	"golang.org/x/crypto/openpgp"
+	"golang.org/x/crypto/openpgp/armor"
+	"golang.org/x/crypto/openpgp/packet"
+	iam "google.golang.org/api/iam/v1"
+	"io"
+	"io/ioutil"
+	"log"
+	"net/http"
+	"os"
+	"regexp"
+	"strings"
+)
+
+type encryptRequest struct {
+	Email string `json:"email"`
+}
+
+type encryptResponse struct {
+	EncryptedKey string `json:"encryptedKey"`
+}
+
+// Initialize this so the background context persists between invocations.
+var iamService *iam.Service
+
+func init() {
+	ctx := context.Background()
+	var err error
+	if iamService, err = iam.NewService(ctx); err != nil {
+		log.Fatal(err)
+	}
+}
+
+func GenerateAndEncrypt(w http.ResponseWriter, r *http.Request) {
+	// Read request bytes
+	data, err := ioutil.ReadAll(r.Body)
+	if err != nil {
+		e := fmt.Sprintf("Error reading request: %v", err)
+		http.Error(w, e, http.StatusBadRequest)
+		return
+	}
+
+	// Parse the request body to get the Service Account email
+	input := encryptRequest{}
+	if err := json.Unmarshal(data, &input); err != nil {
+		e := fmt.Sprintf("Error parsing request: %v", err)
+		http.Error(w, e, http.StatusBadRequest)
+		return
+	}
+
+	// Read in public key
+	recipient, err := readEntity(os.Getenv("PUBLIC_KEY"))
+	if err != nil {
+		e := fmt.Sprintf("Could not parse public key: %v", err)
+		http.Error(w, e, http.StatusInternalServerError)
+		return
+	}
+
+	// Create service account key
+	b64key, err := createServiceAccountKey(input.Email)
+	if err != nil {
+		e := fmt.Sprintf("Could not create Service Account key: %v", err)
+		http.Error(w, e, http.StatusInternalServerError)
+		return
+	}
+	key, err := base64.URLEncoding.DecodeString(b64key)
+	if err != nil {
```

Comment:
Should we actually decode first? I'm wondering if that should happen on the client side (post-decryption).

Reply:
I thought it might avoid encoding issues on different platforms if we base64 before responding. Any reason why this wouldn't work?

Reply:
That part is fine, I'm wondering why we bother to decode before re-encoding (ie. why don't we just encrypt the base64?).

Reply:
This seemed like a better UX thing. Currently the script would be:

```
call-function | base64 --decode > file.gpg
cat file.gpg | gpg --decrypt
```

If we didn't decode and re-encode the user would have to base64 decode the file _after_ it's decrypted, which seemed confusing to me. I think a `.gpg` file should just be encrypted and the user shouldn't have to figure out that it's also encoded

Reply:
Got it, that makes sense!